### PR TITLE
[GUI] [Skins] Add locale InfoLabels for Skins

### PR DESCRIPTION
**Pull request by DeltaMikeCharlie**
## Description
Add the following $INFO labels to extract some locale-based information for use in skins.

$INFO[System.Locale(timezonecountry)] = Kodi settings ‘locale.timezonecountry’.
$INFO[System.Locale(timezone)] = Kodi settings ‘locale.timezone’.
$INFO[System.Locale(region)] = Kodi language system ‘GetCurrentRegion()’.
$INFO[System.Locale(iso)] = Kodi language system ‘GetRegionLocale()’.

## Motivation and context
Ability to adapt skins to regional/locale variations.

## How has this been tested?
A sample skin was updated and the results compared to source locations such as langinfo.xml and GUI settings screens.

## What is the effect on users?
Skin developers will be able to employ locale-specific resources and other customisations where required.

## Screenshots (if appropriate):
Skin
Settings

## Types of change
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [X] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/guilib/guiinfo/SystemGUIInfo.cpp [hunk 1]
```diff
@@ -326,6 +326,31 @@ bool CSystemGUIInfo::GetLabel(std::string& value, const CFileItem *item, int con
 
       return true;
     }
+
+    case SYSTEM_LOCALE_TIMEZONECOUNTRY:
+    {
+        value = CServiceBroker::GetSettingsComponent()->GetSettings()->GetString(CSettings::SETTING_LOCALE_TIMEZONECOUNTRY);
```
Comment (ksooo) [inline xbmc/guilib/guiinfo/SystemGUIInfo.cpp:332]: Indentation looks wrong. Not only here. We use 2 blanks, not 4

**Review by ksooo — COMMENTED** [review]
Nice, needs a bit of love, though.

File: xbmc/GUIInfoManager.cpp [hunk 2]
```diff
@@ -1741,6 +1741,17 @@ const infomap weather[] =        {{ "isfetched",        WEATHER_IS_FETCHED },
 ///     @return the current language.
 ///     <p>
 ///   }
+///   \table_row3{   <b>`System.Locate(type)`</b>,
+///                  \anchor System_Locale
```
Comment (enen92) [inline xbmc/GUIInfoManager.cpp:1745]: This misses revision tags, please do something similar to https://github.com/xbmc/xbmc/blob/master/xbmc/GUIInfoManager.cpp#L149-L151

File: xbmc/GUIInfoManager.cpp [hunk 3]
```diff
@@ -1741,6 +1741,17 @@ const infomap weather[] =        {{ "isfetched",        WEATHER_IS_FETCHED },
 ///     @return the current language.
 ///     <p>
 ///   }
+///   \table_row3{   <b>`System.Locate(type)`</b>,
```
[the remark this answers is not in the document]
Comment (DeltaMikeCharlie) [reply xbmc/GUIInfoManager.cpp:1744]: Well spotted.  Thank you.

File: xbmc/GUIInfoManager.cpp [hunk 4]
```diff
@@ -1741,6 +1741,17 @@ const infomap weather[] =        {{ "isfetched",        WEATHER_IS_FETCHED },
...
+///                  \anchor System_Locale
+///                  _string_,
+///     @return Locale-specific information depending on the requested type.
+///     @param type - Can be one of the following:
+///       - <b>timezonecountry</b> The country name for the current time zone. (Settings -> Interface -> Regional -> Timezone country)
```
[the remark this answers is not in the document]
Comment (DeltaMikeCharlie) [reply xbmc/GUIInfoManager.cpp:1749]: References to settings screen path removed.

File: xbmc/GUIInfoManager.cpp [hunk 5]
```diff
@@ -10138,7 +10149,28 @@ int CGUIInfoManager::TranslateSingleString(const std::string &strCondition, bool
         }
         else if (prop.name == "idletime")
           return AddMultiInfo(CGUIInfo(SYSTEM_IDLE_TIME, atoi(param.c_str())));
-      }
+
```
[the remark this answers is not in the document]
Comment (DeltaMikeCharlie) [reply xbmc/GUIInfoManager.cpp:10152]: Done.

File: xbmc/GUIInfoManager.cpp [hunk 6]
```diff
@@ -10138,7 +10149,28 @@ int CGUIInfoManager::TranslateSingleString(const std::string &strCondition, bool
...
+            }
+            else if (param == "region")
+            {
+                return SYSTEM_LOCALE_REGION;
+            }
+            else if (param == "iso")
+            {
+                return SYSTEM_LOCALE;
+            }
+        }//END of 'system.locale'
```
[the remark this answers is not in the document]
Comment (DeltaMikeCharlie) [reply xbmc/GUIInfoManager.cpp:10171]: Done.

File: xbmc/GUIInfoManager.cpp [hunk 7]
```diff
@@ -10138,7 +10149,28 @@ int CGUIInfoManager::TranslateSingleString(const std::string &strCondition, bool
...
+
+      }//END only 1 parameter
```
[the remark this answers is not in the document]
Comment (DeltaMikeCharlie) [reply xbmc/GUIInfoManager.cpp:10173]: Done.

File: xbmc/GUIInfoManager.cpp [hunk 8]
```diff
@@ -10169,7 +10201,7 @@ int CGUIInfoManager::TranslateSingleString(const std::string &strCondition, bool
         else
           return AddMultiInfo(CGUIInfo(SYSTEM_TIME, StringUtils::TimeStringToSeconds(prop.param(0)), StringUtils::TimeStringToSeconds(prop.param(1))));
       }
-    }
+    }//END 'system'
```
[the remark this answers is not in the document]
Comment (DeltaMikeCharlie) [reply xbmc/GUIInfoManager.cpp:10204]: Done.

File: xbmc/guilib/guiinfo/SystemGUIInfo.cpp [hunk 9]
```diff
@@ -326,6 +326,31 @@ bool CSystemGUIInfo::GetLabel(std::string& value, const CFileItem *item, int con
...
+        value = g_langInfo.GetCurrentRegion();
+        return true;
+    }
+
+    case SYSTEM_LOCALE:
+    {
+        value = g_langInfo.GetRegionLocale();
+        return true;
+    }
+
```
[the remark this answers is not in the document]
Comment (DeltaMikeCharlie) [reply xbmc/guilib/guiinfo/SystemGUIInfo.cpp:353]: Done.

File: xbmc/guilib/guiinfo/SystemGUIInfo.cpp (hunk 1 above)
Comment (DeltaMikeCharlie) [reply xbmc/guilib/guiinfo/SystemGUIInfo.cpp:332]: I applied the Jenkins diff.  That appears to have fixed it.  If not, I can revisit this.

File: xbmc/guilib/guiinfo/GUIInfoLabels.h [hunk 10]
```diff
@@ -564,6 +564,11 @@
 #define SYSTEM_BUILD_VERSION_CODE 1007
 #define SYSTEM_BUILD_VERSION_GIT 1008
 
+#define SYSTEM_LOCALE_TIMEZONECOUNTRY 1009
+#define SYSTEM_LOCALE_TIMEZONE 1010
+#define SYSTEM_LOCALE_REGION 1011
+#define SYSTEM_LOCALE 1012
```
[the remark this answers is not in the document]
Comment (DeltaMikeCharlie) [reply xbmc/guilib/guiinfo/GUIInfoLabels.h:570]: Done.

File: xbmc/GUIInfoManager.cpp [hunk 11]
```diff
@@ -1741,6 +1741,19 @@ const infomap weather[] =        {{ "isfetched",        WEATHER_IS_FETCHED },
...
+///                  _string_,
+///     @return Locale-specific information depending on the requested type.
+///     @param type - Can be one of the following:
+///       - <b>timezonecountry</b> The country name for the current time zone.
+///       - <b>timezone</b> The full timezone name with country and optional region.
+///       - <b>region</b> The currently selected region name within the selected language ( \link System_Language `System.Language` \endlink).
+///       - <b>iso</b> The country code of the currently selected region as specified in <a href="https://kodi.wiki/view/Language_support#What_is_langinfo.xml">langinfo.xml</a>.
+///     <p><hr>
+///     @skinning_v21 **[New Infolabel]** \link System_Locale
+///     `System.Locale(type)`\endlink <p>
```
Comment (enen92) [inline xbmc/GUIInfoManager.cpp:1755]: It'd be nice if you could place the paragraph on a new line

Comment (DeltaMikeCharlie) [reply xbmc/GUIInfoManager.cpp:1755]: @enen92 - I'm not sure exactly what you mean.  Where exactly do you want me to insert the new line?

Comment (enen92) [reply xbmc/GUIInfoManager.cpp:1755]: Just remove the `<p>` that is inline and place it in a new line

Comment (DeltaMikeCharlie) [reply xbmc/GUIInfoManager.cpp]: Done.

File: xbmc/guilib/guiinfo/GUIInfoLabels.h [hunk 12]
```diff
@@ -564,6 +564,11 @@
 #define SYSTEM_BUILD_VERSION_CODE 1007
 #define SYSTEM_BUILD_VERSION_GIT 1008
 
+static constexpr int SYSTEM_LOCALE_TIMEZONECOUNTRY = 1009;
```
Comment (garbear) [inline xbmc/guilib/guiinfo/GUIInfoLabels.h:567]: Unsigned int would probably be more appropriate? These don't go negative (nor even to 0)

Comment (DeltaMikeCharlie) [reply xbmc/guilib/guiinfo/GUIInfoLabels.h:567]: @ksooo specifically requested a type int, but I'm happy to change to unsigned int if required.

Comment (garbear) [reply xbmc/guilib/guiinfo/GUIInfoLabels.h:567]: Required? No, but I appreciate that our team brings different opinions to the table and isn't afraid to challenge existing assumptions. @ksooo what do you think?

Comment (ksooo) [reply xbmc/guilib/guiinfo/GUIInfoLabels.h:567]: @garbear you are right, `unsigned int` is a better fit.

Comment (DeltaMikeCharlie) [reply xbmc/guilib/guiinfo/GUIInfoLabels.h]: Done.

**Review by ksooo — APPROVED** [review]
LGTM. Thanks.

**Comment by DeltaMikeCharlie** [comment]
I think that the latest Jenkins comment is incorrect.  I tried to apply the latest change and git says:
```
error: patch failed: xbmc/GUIInfoManager.cpp:10171
error: xbmc/GUIInfoManager.cpp: patch does not apply
```
It appears that the requested change had already been made so I am ignoring it.

**Comment by ksooo** [comment]
jenkins build this please

**Review by enen92 — APPROVED** [review]
Thanks for your contribution and for taking the comments into consideration. Appreciated.

**Comment by jjd-uk** [comment]
> I think that the latest Jenkins comment is incorrect. I tried to apply the latest change and git says:
> 
> ```
> error: patch failed: xbmc/GUIInfoManager.cpp:10171
> error: xbmc/GUIInfoManager.cpp: patch does not apply
> ```
> 
> It appears that the requested change had already been made so I am ignoring it.

Thanks for your contribution.

I am not sure if it was the case here. but Jenkins can sometimes fail if you do multiple force pushes too close together (you seem to have done 2 force pushes an hour apart), since after the first force push Jenkins will start to build, and if you do another force push before those builds are completed, then that can confuse Jenkins.

Just pointing this out in case you want to make any further contributions.

**Comment by enen92** [comment]
@DeltaMikeCharlie if not asking for too much can you please bring the news in https://forum.kodi.tv/showthread.php?tid=372280? We use that as primary communication vehicle to skinners

**Comment by DeltaMikeCharlie** [comment]
@enen92 - This has been done.

**Comment by DeltaMikeCharlie** [comment]
@ksooo, @jjd-uk, @enen92, @garbear - Thanks for your assistance, comments and feedback.

**Pull request merged**